impl_core/src/main/java/io/opencensus/implcore/metrics/DoubleGaugeMetricImpl.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.implcore.metrics;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.util.concurrent.AtomicDouble;
+import io.opencensus.common.Clock;
+import io.opencensus.common.ToDoubleFunction;
+import io.opencensus.metrics.DoubleGaugeMetric;
+import io.opencensus.metrics.LabelKey;
+import io.opencensus.metrics.LabelValue;
+import io.opencensus.metrics.export.Metric;
+import io.opencensus.metrics.export.MetricDescriptor;
+import io.opencensus.metrics.export.MetricDescriptor.Type;
+import io.opencensus.metrics.export.TimeSeries;
+import io.opencensus.metrics.export.Value;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.annotation.Nullable;
+
+/** Implementation of {@link DoubleGaugeMetric}. */
+public final class DoubleGaugeMetricImpl extends DoubleGaugeMetric implements Meter {
+  private final MetricDescriptor metricDescriptor;
+  private final ConcurrentMap<List<LabelValue>, Point> pointConcurrentHashMap =
+      new ConcurrentHashMap<List<LabelValue>, Point>();
+  private final List<LabelKey> labelKeys;
+
+  DoubleGaugeMetricImpl(String name, String description, String unit, List<LabelKey> labelKeys) {
+    this.metricDescriptor =
+        MetricDescriptor.create(name, description, unit, Type.GAUGE_DOUBLE, labelKeys);
+    this.labelKeys = labelKeys;
+  }
+
+  @Override
+  public Point addPoint(List<LabelValue> labelValues) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+    checkArgument(labelKeys.size() == labelValues.size(), "Incorrect number of labels.");
+
+    return addAndGetPoint(labelValues, null, null);
+  }
+
+  @Override
+  public <T> void addPoint(
+      List<LabelValue> labelValues, @Nullable T obj, ToDoubleFunction<T> function) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+    checkArgument(labelKeys.size() == labelValues.size(), "Incorrect number of labels.");
+
+    addAndGetPoint(labelValues, obj, checkNotNull(function, "function"));
```

Comment:
You need to make a copy of the list if you use the list as the key because the List is not immutable. Same in the Long.